deeplearning/ml4pl/poj104/dataset.py

```diff
@@ -24,56 +30,84 @@
 REPO_ROOT = Path(REPO_ROOT)
 
 
-def load(file):
+# The vocabulary file used in the dataflow experiments.
+PROGRAML_VOCABULARY = REPO_ROOT / "deeplearning/ml4pl/poj104/programl_vocabulary.csv"
+assert PROGRAML_VOCABULARY.is_file()
+
+
+def load(file: str) -> ProgramGraph:
+    """Read a ProgramGraph protocol buffer from file."""
+    graph = ProgramGraph()
     with open(file, 'rb') as f:
-        try:
-            data = pickle.load(f)
-        except EOFError as e:
-            print(f"Failing on {str(file)}")
-            raise e
-    return data
+        graph.ParseFromString(f.read())
+    return graph
+
+
+def load_vocabulary(path: Path):
+    """Read the vocabulary file used in the dataflow experiments."""
+    vocab = {}
+    with open(path) as f:
+        vocab_file = csv.reader(f.readlines(), delimiter="\t")
+        for i, row in enumerate(vocab_file, start=-1):
+            if i == -1:  # Skip the header.
+                continue
+            (_, _, _, text) = row
+            vocab[text] = i
+
+    return vocab
 
 
-def nx2data(nx_graph, class_label=None, ignore_profile_info=True):
-    r"""Converts a :obj:`networkx.Graph` or :obj:`networkx.DiGraph` to a
+class AblationVocab(enum.IntEnum):
+    # No ablation - use the full vocabulary (default).
+    NONE = 0
+    # Ignore the vocabulary - every node has an x value of 0.
+    NO_VOCAB = 1
+    # Use a 3-element vocabulary based on the node type:
+    #    0 - Instruction node
+    #    1 - Variable node
+    #    2 - Constant node
+    NODE_TYPE_ONLY = 2
```

Comment:
Note to self:
the ablation state is part of the cached datasets, but the filenames in the cache don't represent that. This could lead to unwanted mixups.